File: typescript/src/shared/approval-flow/associate.functions.ts
import {z} from 'zod';
import {ApiRoot, ApprovalFlowUpdateAction} from '@commercetools/platform-sdk';
import {SDKError} from '../errors/sdkError';
import {CommercetoolsFuncContext} from '../../types/configuration';
import {
  readApprovalFlowById,
  queryApprovalFlows,
  updateApprovalFlowById,
} from './base.functions';
import {
  readApprovalFlowParameters,
  updateApprovalFlowParameters,
} from './parameters';

export const readApprovalFlow = async (
  apiRoot: ApiRoot,
  context: CommercetoolsFuncContext,
  params: z.infer<typeof readApprovalFlowParameters>
) => {
  if (!context.customerId) {
    throw new Error(
      'Associate ID (customerId) is required for associate operations'
    );
  }
  if (!context.businessUnitKey) {
    throw new Error('Business Unit key is required for associate operations');
  }

  try {
    if (params.id) {
      return await readApprovalFlowById(
        apiRoot,
        context.projectKey,
        context.customerId,
        context.businessUnitKey,
        params.id,
        params.expand
      );
    }

    return await queryApprovalFlows(
      apiRoot,
      context.projectKey,
      context.customerId,
      context.businessUnitKey,
      params.where,
      params.limit,
      params.offset,
      params.sort,
      params.expand
    );
  } catch (error: any) {
    throw new SDKError('Failed to read approval flow as associate', error);
  }
};

export const updateApprovalFlow = async (
  apiRoot: ApiRoot,
  context: CommercetoolsFuncContext,
  params: z.infer<typeof updateApprovalFlowParameters>
) => {
  if (!context.customerId) {
    throw new Error(
      'Associate ID (customerId) is required for associate operations'
    );
  }
  if (!context.businessUnitKey) {
    throw new Error('Business Unit key is required for associate operations');
  }

  try {
    return await updateApprovalFlowById(
      apiRoot,
      context.projectKey,
      context.customerId,
      context.businessUnitKey,
      params.id,
      params.version,
      params.actions as ApprovalFlowUpdateAction[]
    );
  } catch (error: any) {
    throw new SDKError('Failed to update approval flow as associate', error);
  }
};

File: python/commerce_mcp/tools/approval_flow/associate_functions.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any
from .schemas import ReadApprovalFlowParams, UpdateApprovalFlowParams
from ...shared.errors import SDKError
from ...shared.transform import transform_tool_output

if TYPE_CHECKING:
    from ...api import CommercetoolsAPI
    from ...config import CTContext


def _prefix(ctx: "CTContext") -> str:
    return (
        f"/as-associate/{ctx.customer_id}"
        f"/in-business-unit/key={ctx.business_unit_key}"
    )


async def read_approval_flow(
    params: ReadApprovalFlowParams,
    api: "CommercetoolsAPI",
    ctx: "CTContext",
) -> str:
    try:
        prefix = _prefix(ctx)
        if params.id:
            query: dict[str, Any] = {}
            if params.expand:
                query["expand"] = params.expand
            result = await api.get(f"{prefix}/approval-flows/{params.id}", params=query or None)
            return transform_tool_output(result)
        query = {"limit": params.limit or 10}
        if params.where:
            query["where"] = params.where
        if params.offset is not None:
            query["offset"] = params.offset
        if params.sort:
            query["sort"] = params.sort
        if params.expand:
            query["expand"] = params.expand
        result = await api.get(f"{prefix}/approval-flows", params=query)
        return transform_tool_output(result)
    except Exception as e:
        raise SDKError("read approval flow as associate", e)


async def update_approval_flow(
    params: UpdateApprovalFlowParams,
    api: "CommercetoolsAPI",
    ctx: "CTContext",
) -> str:
    try:
        prefix = _prefix(ctx)
        body: dict[str, Any] = {"version": params.version, "actions": params.actions}
        result = await api.post(f"{prefix}/approval-flows/{params.id}", body=body)
        return transform_tool_output(result)
    except Exception as e:
        raise SDKError("update approval flow as associate", e)
